File: nmpolicy/internal/resolver/resolver.go
/*
 * Copyright 2021 NMPolicy Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 *	  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package resolver

import (
	"fmt"

	"sigs.k8s.io/yaml"

	"github.com/nmstate/nmpolicy/nmpolicy/internal/ast"
	"github.com/nmstate/nmpolicy/nmpolicy/types"
)

type resolver struct {
	currentState   map[string]interface{}
	capturedStates map[string]map[string]interface{}
	captureASTPool map[string]ast.Node
}

type Resolver struct{}

func New() Resolver {
	return Resolver{}
}

func (Resolver) Resolve(captureASTPool map[string]ast.Node, currentState []byte) (Result, error) {
	return newWithCaptureASTPool(captureASTPool).Resolve(currentState)
}

func (Resolver) ResolveCaptureEntryPath(captureEntryPathAST ast.Node,
	capturedStates map[string]map[string]interface{}) (interface{}, error) {
	return newWithCapturedStates(capturedStates).ResolveCaptureEntryPath(captureEntryPathAST)
}

func newEmpty() *resolver {
	return &resolver{
		currentState:   map[string]interface{}{},
		capturedStates: map[string]map[string]interface{}{},
		captureASTPool: map[string]ast.Node{},
	}
}

func newWithCaptureASTPool(captureASTPool map[string]ast.Node) *resolver {
	r := newEmpty()
	r.captureASTPool = captureASTPool
	return r
}

func newWithCapturedStates(capturedStates map[string]map[string]interface{}) *resolver {
	r := newEmpty()
	r.capturedStates = capturedStates
	return r
}

func (r *resolver) Resolve(currentState []byte) (Result, error) {
	err := yaml.Unmarshal(currentState, &r.currentState)
	if err != nil {
		return Result{}, wrapWithResolveError(err)
	}

	result := Result{
		Marshaled:   map[string]types.CapturedState{},
		Unmarshaled: map[string]map[string]interface{}{},
	}
	for captureEntryName := range r.captureASTPool {
		capturedStateEntry, err := r.resolveCaptureEntryName(captureEntryName)
		if err != nil {
			return Result{}, wrapWithResolveError(err)
		}
		marshaledCapturedStateEntry, err := yaml.Marshal(capturedStateEntry)
		if err != nil {
			return Result{}, wrapWithResolveError(err)
		}
		result.Marshaled[captureEntryName] = types.CapturedState{
			State:    marshaledCapturedStateEntry,
			MetaInfo: types.MetaInfo{},
		}
		result.Unmarshaled[captureEntryName] = capturedStateEntry
	}
	return result, nil
}

func (r *resolver) ResolveCaptureEntryPath(captureEntryPath ast.Node) (interface{}, error) {
	return r.resolveCaptureEntryPath(captureEntryPath)
}

func (r *resolver) resolveCaptureEntryName(captureEntryName string) (map[string]interface{}, error) {
	capturedStateEntry, ok := r.capturedStates[captureEntryName]
	if ok {
		return capturedStateEntry, nil
	}
	captureASTEntry, ok := r.captureASTPool[captureEntryName]
	if !ok {
		return nil, fmt.Errorf("capture entry '%s' not found", captureEntryName)
	}
	capturedStateEntry, err := r.resolveCaptureASTEntry(captureASTEntry)
	if err != nil {
		return nil, err
	}
	r.capturedStates[captureEntryName] = capturedStateEntry
	return capturedStateEntry, nil
}

func (r resolver) resolveCaptureASTEntry(captureASTEntry ast.Node) (map[string]interface{}, error) {
	if captureASTEntry.EqFilter != nil {
		return r.resolveEqFilter(captureASTEntry.EqFilter)
	}
	return nil, fmt.Errorf("root node has unsupported operation : %v", captureASTEntry)
}

func (r resolver) resolveEqFilter(operator *ast.TernaryOperator) (map[string]interface{}, error) {
	inputSource, err := r.resolveInputSource((*operator)[0], r.currentState)
	if err != nil {
		return nil, wrapWithEqFilterError(err)
	}

	path, err := r.resolvePath((*operator)[1])
	if err != nil {
		return nil, wrapWithEqFilterError(err)
	}
	filteredValue, err := r.resolveFilteredValue((*operator)[2])
	if err != nil {
		return nil, wrapWithEqFilterError(err)
	}
	filteredState, err := filter(inputSource, path.steps, *filteredValue)
	if err != nil {
		return nil, wrapWithEqFilterError(err)
	}
	return filteredState, nil
}

func (r resolver) resolveInputSource(inputSourceNode ast.Node, currentState map[string]interface{}) (map[string]interface{}, error) {
	if ast.CurrentStateIdentity().DeepEqual(inputSourceNode.Terminal) {
		return currentState, nil
	}

	return nil, fmt.Errorf("not supported input source %v. Only the current state is supported", inputSourceNode)
}

func (r resolver) resolveFilteredValue(filteredValueNode ast.Node) (*ast.Node, error) {
	if filteredValueNode.String != nil {
		return &filteredValueNode, nil
	} else if filteredValueNode.Path != nil {
		resolvedCaptureEntryPath, err := r.resolveCaptureEntryPath(filteredValueNode)
		if err != nil {
			return nil, err
		}
		terminal, err := newTerminalFromInterface(resolvedCaptureEntryPath)
		if err != nil {
			return nil, err
		}
		return &ast.Node{
			Terminal: *terminal,
		}, nil
	} else {
		return nil, fmt.Errorf("not supported filtered value. Only string or paths are supported")
	}
}

func (r resolver) resolveCaptureEntryPath(pathNode ast.Node) (interface{}, error) {
	resolvedPath, err := r.resolvePath(pathNode)
	if err != nil {
		return nil, err
	}
	if resolvedPath.captureEntryName == "" {
		return nil, fmt.Errorf("not supported filtered value path. Only paths with a capture entry reference are supported")
	}
	capturedStateEntry, err := r.resolveCaptureEntryName(resolvedPath.captureEntryName)
	if err != nil {
		return nil, err
	}
	return resolvedPath.walkState(capturedStateEntry)
}

func (r resolver) resolvePath(pathNode ast.Node) (*captureEntryNameAndSteps, error) {
	if pathNode.Path == nil {
		return nil, fmt.Errorf("invalid path type %T", pathNode)
	} else if len(*pathNode.Path) == 0 {
		return nil, fmt.Errorf("empty path length")
	} else if (*pathNode.Path)[0].Identity == nil {
		return nil, fmt.Errorf("path first step has to be an identity")
	}
	resolvedPath := captureEntryNameAndSteps{
		steps: *pathNode.Path,
	}
	if *resolvedPath.steps[0].Identity == "capture" {
		const captureRefSize = 2
		if len(resolvedPath.steps) < captureRefSize || resolvedPath.steps[1].Identity == nil {
			return nil, fmt.Errorf("path capture ref is missing capture entry name")
		}
		resolvedPath.captureEntryName = *resolvedPath.steps[1].Identity
		if len(resolvedPath.steps) > captureRefSize {
			resolvedPath.steps = resolvedPath.steps[2:len(resolvedPath.steps)]
		}
	}
	return &resolvedPath, nil
}

File: rust/src/lib/resolve/resolver.rs
use crate::{
    ast::node::{current_state_identity, Node, NodeKind, TernaryOperator},
    error::{evaluation_error, NmpolicyError},
    resolve::{filter, path::Path, replace, walk},
    types::{Capture, CapturedState, CapturedStates, NMState},
};

use serde_json::Value;

pub(crate) struct Resolver {
    capture: Capture,
    current_state: Option<NMState>,
    captured_states: CapturedStates,
    current_node: Option<Box<Node>>,
    current_expression: Option<String>,
}

impl Resolver {
    pub(crate) fn new(capture: Capture) -> Self {
        Self {
            capture,
            current_state: None,
            captured_states: CapturedStates::new(),
            current_node: None,
            current_expression: None,
        }
    }

    pub(crate) fn resolve(
        &mut self,
        current_state: NMState,
        cache: Option<CapturedStates>,
    ) -> Result<CapturedStates, NmpolicyError> {
        self.current_state = Some(current_state);
        if let Some(c) = cache {
            self.captured_states = c
        };
        for capture_entry_name in self.capture.clone().keys() {
            if let Err(mut e) = self.resolve_capture_entry_by_name(capture_entry_name) {
                e = e.expression(self.current_expression.clone()).resolver();
                match self.current_node.clone() {
                    Some(current_node) => return Err(e.pos(current_node.pos)),
                    None => return Err(e),
                }
            }
        }

        Ok(self.captured_states.clone())
    }

    fn resolve_capture_entry_by_name(
        &mut self,
        capture_entry_name: &String,
    ) -> Result<NMState, NmpolicyError> {
        if let Some(captured_state_entry) = self.captured_states.get(capture_entry_name) {
            return Ok(captured_state_entry.clone().state);
        }
        match self.capture.get(capture_entry_name) {
            Some(capture_entry) => {
                self.current_expression = Some(capture_entry.expression.clone());
                self.current_node = Some(capture_entry.ast.clone());
            }
            _ => {
                return Err(evaluation_error(format!(
                    "capture entry '{}' not found",
                    capture_entry_name.clone()
                )));
            }
        }
        let resolved_state = self.resolve_current_capture_entry()?;
        self.captured_states.insert(
            capture_entry_name.to_string(),
            CapturedState {
                state: resolved_state.clone(),
                meta_info: None,
            },
        );
        Ok(resolved_state)
    }

    fn resolve_current_capture_entry(&mut self) -> Result<NMState, NmpolicyError> {
        match self.current_node.clone() {
            Some(current_node) => match current_node.kind {
                NodeKind::EqFilter(lhs, ms, rhs) => match self.resolve_eqfilter((lhs, ms, rhs)) {
                    Ok(state) => Ok(state),
                    Err(e) => Err(e.eqfilter()),
                },
                NodeKind::Replace(lhs, ms, rhs) => match self.resolve_replace((lhs, ms, rhs)) {
                    Ok(state) => Ok(state),
                    Err(e) => Err(e.replace()),
                },
                NodeKind::Path(_) => self.resolve_path_filter(*current_node),
                _ => Err(evaluation_error(format!(
                    "root node has unsupported operation : {current_node}"
                ))),
            },
            None => Ok(NMState::new()),
        }
    }

    fn resolve_eqfilter(&mut self, operator: TernaryOperator) -> Result<NMState, NmpolicyError> {
        let (input_source, path, value) = self.resolve_ternary_operator(operator)?;
        filter::visit_state(input_source, path, value)
    }

    fn resolve_replace(&mut self, operator: TernaryOperator) -> Result<NMState, NmpolicyError> {
        let (input_source, path, value) = self.resolve_ternary_operator(operator)?;
        replace::visit_state(input_source, path, value)
    }

    fn resolve_path_filter(&mut self, node: Node) -> Result<NMState, NmpolicyError> {
        let path = Path::compose_from_node(node)?;
        let current_state = self.current_state.clone().unwrap();
        filter::visit_state(current_state, path, Value::Null)
    }
    fn resolve_ternary_operator(
        &mut self,
        operator: TernaryOperator,
    ) -> Result<(NMState, Path, Value), NmpolicyError> {
        let operator_node = self.current_node.clone();
        let (lhs, ms, rhs) = operator;

        self.current_node = Some(lhs);
        let input_source = self.resolve_input_source()?;

        self.current_node = Some(ms.clone());
        let path = Path::compose_from_node(*ms)?;

        self.current_node = Some(rhs.clone());
        let value: Value = match rhs.kind {
            NodeKind::Str(string) => Value::String(string),
            NodeKind::Path(_) => self.resolve_capture_entry_path()?,
            _ => {
                return Err(evaluation_error(
                    "not supported value. Only string or capture entry path are supported"
                        .to_string(),
                ));
            }
        };
        self.current_node = operator_node;
        Ok((input_source, path, value))
    }

    fn resolve_capture_entry_path(&mut self) -> Result<Value, NmpolicyError> {
        let current_node = self.current_node.clone().unwrap();
        let path = Path::compose_from_node(*current_node)?;
        match path.clone().capture_entry_name {
            Some(capture_entry_name) => {
                let captured_state_entry = self.resolve_capture_entry_by_name(&capture_entry_name)?;
                walk::visit_state(captured_state_entry, path)
            }
            None => Err(evaluation_error("not supported filtered value path. Only paths with a capture entry reference are supported".to_string()))
        }
    }

    fn resolve_input_source(&mut self) -> Result<NMState, NmpolicyError> {
        let current_node = self.current_node.clone().unwrap();
        match current_node.kind {
            n if n == current_state_identity() => Ok(self.current_state.clone().unwrap()),
            _ => match Path::compose_from_node(*current_node.clone()) {
                Ok(path) => match path.capture_entry_name {
                    Some(capture_entry_name) => {
                        self.resolve_capture_entry_by_name(&capture_entry_name)
                    }
                    None => {
                        return Err(evaluation_error(format!(
                            "invalid path input source ({}), only capture reference is supported",
                            current_node
                        )))
                    }
                },
                Err(_) => {
                    return Err(evaluation_error(format!("invalid input source ({}), only current state or capture reference is supported", current_node)));
                }
            },
        }
    }
}
